File: Dev/TemplateFactory/tools/check_validation_contract.js
#!/usr/bin/env node

const { execFileSync } = require("child_process");
const { repoPath } = require("./node_utils");

const ROOT = process.cwd();
const CONTRACT = "Dev/TemplateFactory/modules/validation_contract.yaml";
const FIELDS_CORE = "Dev/TemplateFactory/modules/fields_core.yaml";
const FRONTMATTER_PROFILES = "Dev/TemplateFactory/modules/frontmatter_profiles.yaml";
const RUNTIME_PROFILES = "Dev/TemplateFactory/modules/runtime_profiles.yaml";

function loadYamlModule(relPath) {
    const script = [
        "import json, sys, yaml",
        "with open(sys.argv[1], encoding='utf-8') as handle:",
        "    data = yaml.safe_load(handle) or {}",
        "print(json.dumps(data, ensure_ascii=False))"
    ].join("\n");
    const stdout = execFileSync("python3", ["-c", script, repoPath(ROOT, relPath)], {
        encoding: "utf8",
        maxBuffer: 4 * 1024 * 1024
    });
    return JSON.parse(stdout);
}

function valuesForCoreField(fieldsCore, fieldName) {
    for (const group of Object.values(fieldsCore.fields ?? {})) {
        for (const field of group ?? []) {
            if (field?.name === fieldName) return new Set((field.values ?? []).map(String));
        }
    }
    return new Set();
}

function addField(knownFields, value) {
    const field = String(value ?? "").trim();
    if (field) knownFields.add(field);
}

function knownFrontmatterFields(fieldsCore, frontmatterProfiles) {
    const knownFields = new Set();
    for (const group of Object.values(fieldsCore.fields ?? {})) {
        for (const field of group ?? []) addField(knownFields, field?.name);
    }

    for (const value of Object.values(frontmatterProfiles.field_catalog ?? {})) {
        if (Array.isArray(value)) {
            for (const field of value) addField(knownFields, field);
            continue;
        }
        if (value && typeof value === "object") {
            for (const fields of Object.values(value)) {
                for (const field of fields ?? []) addField(knownFields, field);
            }
        }
    }

    for (const profile of Object.values(frontmatterProfiles.profiles ?? {})) {
        for (const field of profile.required_fields ?? []) addField(knownFields, field);
        for (const field of profile.sample_values ?? []) addField(knownFields, field);
        for (const field of profile.fields ?? []) {
            addField(knownFields, field?.key);
            addField(knownFields, field?.value);
        }
    }
    return knownFields;
}

function requireNonEmptyArray(errors, value, path) {
    if (!Array.isArray(value) || value.length === 0) {
        errors.push(`${CONTRACT}: ${path} deve essere lista non vuota`);
        return [];
    }
    const normalized = value.map(String).filter(Boolean);
    const scalarOnly = value.every(item => item === null || !["object", "function"].includes(typeof item));
    if (scalarOnly && new Set(normalized).size !== normalized.length) {
        errors.push(`${CONTRACT}: ${path} contiene valori duplicati`);
    }
    return normalized;
}

function requireNonEmptyMap(errors, value, path) {
    if (!value || typeof value !== "object" || Array.isArray(value) || Object.keys(value).length === 0) {
        errors.push(`${CONTRACT}: ${path} deve essere mappa non vuota`);
        return {};
    }
    return value;
}

function requireNumberAtLeast(errors, value, path, minimum) {
    const number = Number(value);
    if (!Number.isFinite(number) || number < minimum) {
        errors.push(`${CONTRACT}: ${path} deve essere numero >= ${minimum}`);
    }
    return number;
}

function requireNonEmptyString(errors, value, path) {
    const text = String(value ?? "").trim();
    if (!text) errors.push(`${CONTRACT}: ${path} deve essere stringa non vuota`);
    return text;
}

function requireKnownField(errors, knownFields, field, path) {
    if (!knownFields.has(field)) {
        errors.push(`${CONTRACT}: ${path} usa campo non dichiarato in fields_core/frontmatter_profiles (${field})`);
    }
}

function requireKnownFieldArray(errors, knownFields, value, path) {
    const fields = requireNonEmptyArray(errors, value, path);
    for (const field of fields) requireKnownField(errors, knownFields, field, path);
    return fields;
}

function requireKnownFieldGroups(errors, knownFields, value, path) {
    if (!Array.isArray(value) || value.length === 0) {
        errors.push(`${CONTRACT}: ${path} deve essere lista non vuota`);
        return [];
    }
    return value.map((group, index) => requireKnownFieldArray(errors, knownFields, group, `${path}.${index}`));
}

function validateCategoryList(errors, categories, allowedCategories, path) {
    for (const category of categories) {
        if (!allowedCategories.has(category)) {
            errors.push(`${CONTRACT}: ${path} contiene categoria non dichiarata in fields_core.yaml (${category})`);
        }
    }
}

function setEquals(left, right) {
    if (left.size !== right.size) return false;
    for (const value of left) {
        if (!right.has(value)) return false;
    }
    return true;
}

const errors = [];
const contract = loadYamlModule(CONTRACT);
const fieldsCore = loadYamlModule(FIELDS_CORE);
const frontmatterProfiles = loadYamlModule(FRONTMATTER_PROFILES);
const runtimeProfiles = loadYamlModule(RUNTIME_PROFILES);
const allowedCategories = valuesForCoreField(fieldsCore, "categoria");
const allowedStates = valuesForCoreField(fieldsCore, "stato");
const knownFields = knownFrontmatterFields(fieldsCore, frontmatterProfiles);

for (const key of ["live_entity_categories", "codex_categories"]) {
    validateCategoryList(errors, requireNonEmptyArray(errors, contract[key], key), allowedCategories, key);
}

const categoryValidation = requireNonEmptyMap(errors, contract.category_validation, "category_validation");
const typeExemptCategories = new Set(requireNonEmptyArray(errors, categoryValidation.type_exempt_categories, "category_validation.type_exempt_categories"));
const requiredFieldExemptCategories = new Set(requireNonEmptyArray(errors, categoryValidation.required_field_exempt_categories, "category_validation.required_field_exempt_categories"));
validateCategoryList(errors, [...typeExemptCategories], allowedCategories, "category_validation.type_exempt_categories");
validateCategoryList(errors, [...requiredFieldExemptCategories], allowedCategories, "category_validation.required_field_exempt_categories");

const allowedTypesByCategory = requireNonEmptyMap(errors, contract.allowed_types_by_category, "allowed_types_by_category");
for (const [category, types] of Object.entries(allowedTypesByCategory)) {
    if (!allowedCategories.has(category)) {
        errors.push(`${CONTRACT}: allowed_types_by_category contiene categoria non dichiarata in fields_core.yaml (${category})`);
    }
    requireNonEmptyArray(errors, types, `allowed_types_by_category.${category}`);
    if (typeExemptCategories.has(category)) {
        errors.push(`${CONTRACT}: ${category} risulta sia in allowed_types_by_category sia in category_validation.type_exempt_categories`);
    }
}
for (const category of allowedCategories) {
    if (!Object.prototype.hasOwnProperty.call(allowedTypesByCategory, category) && !typeExemptCategories.has(category)) {
        errors.push(`${CONTRACT}: allowed_types_by_category non copre categoria ${category} e non la dichiara esente`);
    }
}

const runtimeLocationTypes = new Set((runtimeProfiles.profiles?.luogo?.type_options ?? [])
    .map(option => String(option?.id ?? ""))
    .filter(Boolean));
if (runtimeLocationTypes.size) {
    const contractLocationTypes = new Set((allowedTypesByCategory.luogo ?? []).map(String));
    if (!setEquals(runtimeLocationTypes, contractLocationTypes)) {
        errors.push(`${CONTRACT}: allowed_types_by_category.luogo deve corrispondere a runtime_profiles.luogo.type_options`);
    }
}

const requiredFieldsByCategory = requireNonEmptyMap(errors, contract.required_fields_by_category, "required_fields_by_category");
for (const [category, fields] of Object.entries(requiredFieldsByCategory)) {
    if (!allowedCategories.has(category)) {
        errors.push(`${CONTRACT}: required_fields_by_category contiene categoria non dichiarata in fields_core.yaml (${category})`);
    }
    requireKnownFieldArray(errors, knownFields, fields, `required_fields_by_category.${category}`);
    if (requiredFieldExemptCategories.has(category)) {
        errors.push(`${CONTRACT}: ${category} risulta sia in required_fields_by_category sia in category_validation.required_field_exempt_categories`);
    }
}
for (const category of allowedCategories) {
    if (!Object.prototype.hasOwnProperty.call(requiredFieldsByCategory, category) && !requiredFieldExemptCategories.has(category)) {
        errors.push(`${CONTRACT}: required_fields_by_category non copre categoria ${category} e non la dichiara esente`);
    }
}

const readyRequirements = requireNonEmptyMap(errors, contract.state_ready_requirements, "state_ready_requirements");
const readyStates = requireNonEmptyMap(errors, readyRequirements.pronto, "state_ready_requirements.pronto");
const allowedReadyKeys = new Set([
    ...allowedCategories,
    ...Object.keys(allowedTypesByCategory),
    ...Object.values(allowedTypesByCategory).flat().map(String)
]);
for (const [categoryOrType, requirement] of Object.entries(readyStates)) {
    if (!allowedReadyKeys.has(categoryOrType)) {
        errors.push(`${CONTRACT}: state_ready_requirements.pronto contiene chiave non riconducibile a categoria o tipo (${categoryOrType})`);
    }
    requireKnownFieldArray(errors, knownFields, requirement?.any_of_fields, `state_ready_requirements.pronto.${categoryOrType}.any_of_fields`);
}

requireKnownFieldArray(errors, knownFields, contract.private_frontmatter_fields, "private_frontmatter_fields");
requireNonEmptyArray(errors, contract.private_text_terms, "private_text_terms");

const liveEntityPolicy = requireNonEmptyMap(errors, contract.live_entity_policy, "live_entity_policy");
requireKnownFieldArray(errors, knownFields, liveEntityPolicy.require_any_of_fields, "live_entity_policy.require_any_of_fields");
requireKnownFieldArray(errors, knownFields, liveEntityPolicy.sheet_require_any_of_fields, "live_entity_policy.sheet_require_any_of_fields");

const codexArticlePolicy = requireNonEmptyMap(errors, contract.codex_article_policy, "codex_article_policy");
requireKnownFieldArray(errors, knownFields, codexArticlePolicy.identity_any_of_fields, "codex_article_policy.identity_any_of_fields");
requireKnownFieldArray(errors, knownFields, codexArticlePolicy.table_use_any_of_fields, "codex_article_policy.table_use_any_of_fields");
requireKnownFieldArray(errors, knownFields, codexArticlePolicy.dm_layer_any_of_fields, "codex_article_policy.dm_layer_any_of_fields");
requireKnownFieldArray(errors, knownFields, codexArticlePolicy.operational_link_fields, "codex_article_policy.operational_link_fields");

const sessionPlayabilityPolicy = requireNonEmptyMap(errors, contract.session_playability_policy, "session_playability_policy");
for (const state of requireNonEmptyArray(errors, sessionPlayabilityPolicy.active_states, "session_playability_policy.active_states")) {
    if (!allowedStates.has(state)) {
        errors.push(`${CONTRACT}: session_playability_policy.active_states usa stato non dichiarato in fields_core.yaml (${state})`);
    }
}
requireKnownFieldArray(errors, knownFields, sessionPlayabilityPolicy.playable_any_of_fields, "session_playability_policy.playable_any_of_fields");
requireNumberAtLeast(errors, sessionPlayabilityPolicy.min_world_anchors, "session_playability_policy.min_world_anchors", 1);
const worldAnchorGroups = requireKnownFieldGroups(errors, knownFields, sessionPlayabilityPolicy.world_anchor_groups, "session_playability_policy.world_anchor_groups");
if (worldAnchorGroups.length && Number(sessionPlayabilityPolicy.min_world_anchors) > worldAnchorGroups.length) {
    errors.push(`${CONTRACT}: session_playability_policy.min_world_anchors supera i gruppi world_anchor_groups`);
}
const encounterMaterialField = requireNonEmptyString(errors, sessionPlayabilityPolicy.encounter_material_field, "session_playability_policy.encounter_material_field");
if (encounterMaterialField) requireKnownField(errors, knownFields, encounterMaterialField, "session_playability_policy.encounter_material_field");

const mapReviewPolicy = requireNonEmptyMap(errors, contract.map_review_policy, "map_review_policy");
const playableMapUses = new Set(requireNonEmptyArray(errors, mapReviewPolicy.playable_uses, "map_review_policy.playable_uses"));
requireNonEmptyArray(errors, mapReviewPolicy.structured_uses, "map_review_policy.structured_uses");
requireKnownFieldArray(errors, knownFields, mapReviewPolicy.table_use_any_of_fields, "map_review_policy.table_use_any_of_fields");
requireKnownFieldArray(errors, knownFields, mapReviewPolicy.visibility_any_of_fields, "map_review_policy.visibility_any_of_fields");
requireKnownFieldArray(errors, knownFields, mapReviewPolicy.dm_private_any_of_fields, "map_review_policy.dm_private_any_of_fields");
requireKnownFieldArray(errors, knownFields, mapReviewPolicy.ready_playable_link_any_of_fields, "map_review_policy.ready_playable_link_any_of_fields");
requireKnownFieldArray(errors, knownFields, mapReviewPolicy.ready_structural_link_any_of_fields, "map_review_policy.ready_structural_link_any_of_fields");
const zoomUse = requireNonEmptyString(errors, mapReviewPolicy.zoom_use, "map_review_policy.zoom_use");
if (zoomUse && !playableMapUses.has(zoomUse)) {
    errors.push(`${CONTRACT}: map_review_policy.zoom_use deve essere incluso in map_review_policy.playable_uses`);
}

requireNonEmptyArray(errors, contract.playability_rules, "playability_rules");
const playabilityRules = Array.isArray(contract.playability_rules) ? contract.playability_rules : [];
const ruleIds = new Set();
for (const rule of playabilityRules) {
    const id = String(rule?.id ?? "");
    if (!id) {
        errors.push(`${CONTRACT}: playability_rules contiene regola senza id`);
        continue;
    }
    if (ruleIds.has(id)) errors.push(`${CONTRACT}: playability_rules id duplicato (${id})`);
    ruleIds.add(id);

    if (!rule.warning) errors.push(`${CONTRACT}: playability_rules.${id} senza warning`);
    if (!rule.require_value && !rule.require_any_of) {
        errors.push(`${CONTRACT}: playability_rules.${id} senza require_value o require_any_of`);
    }
    if (rule.require_value) requireKnownField(errors, knownFields, String(rule.require_value), `playability_rules.${id}.require_value`);
    if (rule.require_any_of) requireKnownFieldArray(errors, knownFields, rule.require_any_of, `playability_rules.${id}.require_any_of`);
    if (rule.tipo_in) requireNonEmptyArray(errors, rule.tipo_in, `playability_rules.${id}.tipo_in`);
    if (rule.path_prefixes) requireNonEmptyArray(errors, rule.path_prefixes, `playability_rules.${id}.path_prefixes`);
    if (rule.when_value_present) requireKnownField(errors, knownFields, String(rule.when_value_present), `playability_rules.${id}.when_value_present`);
    if (rule.when_any_of_present) requireKnownFieldArray(errors, knownFields, rule.when_any_of_present, `playability_rules.${id}.when_any_of_present`);
    if (rule.any_field_equals) requireNonEmptyArray(errors, rule.any_field_equals, `playability_rules.${id}.any_field_equals`);
    for (const condition of rule.any_field_equals ?? []) {
        requireKnownField(errors, knownFields, String(condition?.field ?? ""), `playability_rules.${id}.any_field_equals.field`);
    }
    if (rule.number_field_gt) requireKnownField(errors, knownFields, String(rule.number_field_gt.field ?? ""), `playability_rules.${id}.number_field_gt.field`);
    if (rule.number_field_gte) requireKnownField(errors, knownFields, String(rule.number_field_gte.field ?? ""), `playability_rules.${id}.number_field_gte.field`);

    if (rule.categoria && !allowedCategories.has(String(rule.categoria))) {
        errors.push(`${CONTRACT}: playability_rules.${id} usa categoria non dichiarata in fields_core.yaml (${rule.categoria})`);
    }
}

if (errors.length) {
    console.error("Validation contract non valido:");
    for (const error of errors) console.error(`- ${error}`);
    process.exit(1);
}

console.log(`Validation contract OK: ${Object.keys(allowedTypesByCategory).length} categorie tipo, ${Object.keys(requiredFieldsByCategory).length} categorie required, ${Object.keys(readyStates).length} policy pronto, ${playabilityRules.length} regole giocabilita, policy live/sessione/mappe validate.`);

File: Dev/Tools/python/check_validation_contract.py
#!/usr/bin/env python3

from __future__ import annotations

import sys
from typing import Any

from template_factory_utils import ROOT, load_yaml

sys.dont_write_bytecode = True

CONTRACT_PATH = ROOT / "Dev" / "Source" / "YAML" / "canonical" / "validation_contract.yaml"
FIELDS_CORE_PATH = ROOT / "Dev" / "Source" / "YAML" / "canonical" / "fields_core.yaml"
FRONTMATTER_PROFILES_PATH = ROOT / "Dev" / "Source" / "YAML" / "canonical" / "frontmatter_profiles.yaml"
RUNTIME_PROFILES_PATH = ROOT / "Dev" / "Source" / "YAML" / "canonical" / "runtime_profiles.yaml"


def values_for_core_field(fields_core: dict[str, Any], field_name: str) -> set[str]:
    for group in (fields_core.get("fields") or {}).values():
        if not isinstance(group, list):
            continue
        for field in group:
            if isinstance(field, dict) and field.get("name") == field_name:
                return {str(value) for value in field.get("values", [])}
    return set()


def add_field(known_fields: set[str], value: Any) -> None:
    field = str(value or "").strip()
    if field:
        known_fields.add(field)


def known_frontmatter_fields(fields_core: dict[str, Any], frontmatter_profiles: dict[str, Any]) -> set[str]:
    known_fields: set[str] = set()
    for group in (fields_core.get("fields") or {}).values():
        if not isinstance(group, list):
            continue
        for field in group:
            if isinstance(field, dict):
                add_field(known_fields, field.get("name"))

    for value in (frontmatter_profiles.get("field_catalog") or {}).values():
        if isinstance(value, list):
            for field in value:
                add_field(known_fields, field)
            continue
        if isinstance(value, dict):
            for fields in value.values():
                if isinstance(fields, list):
                    for field in fields:
                        add_field(known_fields, field)

    for profile in (frontmatter_profiles.get("profiles") or {}).values():
        if not isinstance(profile, dict):
            continue
        for field in profile.get("required_fields", []) or []:
            add_field(known_fields, field)
        sample_values = profile.get("sample_values", []) or []
        if isinstance(sample_values, list):
            for field in sample_values:
                add_field(known_fields, field)
        for field in profile.get("fields", []) or []:
            if isinstance(field, dict):
                add_field(known_fields, field.get("key"))
                add_field(known_fields, field.get("value"))
    return known_fields


def require_non_empty_array(errors: list[str], value: Any, path: str) -> list[str]:
    if not isinstance(value, list) or not value:
        errors.append(f"{CONTRACT_PATH.relative_to(ROOT)}: {path} deve essere lista non vuota")
        return []
    normalized = [str(item) for item in value if str(item)]
    scalar_only = all(item is None or not isinstance(item, (dict, list, set, tuple)) for item in value)
    if scalar_only and len(set(normalized)) != len(normalized):
        errors.append(f"{CONTRACT_PATH.relative_to(ROOT)}: {path} contiene valori duplicati")
    return normalized


def require_non_empty_map(errors: list[str], value: Any, path: str) -> dict[str, Any]:
    if not isinstance(value, dict) or not value:
        errors.append(f"{CONTRACT_PATH.relative_to(ROOT)}: {path} deve essere mappa non vuota")
        return {}
    return value


def require_number_at_least(errors: list[str], value: Any, path: str, minimum: float) -> float:
    try:
        number = float(value)
    except (TypeError, ValueError):
        number = float("nan")
    if number != number or number < minimum:
        errors.append(f"{CONTRACT_PATH.relative_to(ROOT)}: {path} deve essere numero >= {minimum:g}")
    return number


def require_non_empty_string(errors: list[str], value: Any, path: str) -> str:
    text = str(value or "").strip()
    if not text:
        errors.append(f"{CONTRACT_PATH.relative_to(ROOT)}: {path} deve essere stringa non vuota")
    return text


def require_known_field(errors: list[str], known_fields: set[str], field: str, path: str) -> None:
    if field not in known_fields:
        errors.append(
            f"{CONTRACT_PATH.relative_to(ROOT)}: {path} usa campo non dichiarato in "
            f"fields_core/frontmatter_profiles ({field})"
        )


def require_known_field_array(errors: list[str], known_fields: set[str], value: Any, path: str) -> list[str]:
    fields = require_non_empty_array(errors, value, path)
    for field in fields:
        require_known_field(errors, known_fields, field, path)
    return fields


def require_known_field_groups(errors: list[str], known_fields: set[str], value: Any, path: str) -> list[list[str]]:
    if not isinstance(value, list) or not value:
        errors.append(f"{CONTRACT_PATH.relative_to(ROOT)}: {path} deve essere lista non vuota")
        return []
    return [
        require_known_field_array(errors, known_fields, group, f"{path}.{index}")
        for index, group in enumerate(value)
    ]


def validate_category_list(errors: list[str], categories: list[str], allowed_categories: set[str], path: str) -> None:
    for category in categories:
        if category not in allowed_categories:
            errors.append(
                f"{CONTRACT_PATH.relative_to(ROOT)}: {path} contiene categoria non dichiarata "
                f"in fields_core.yaml ({category})"
            )


def set_equals(left: set[str], right: set[str]) -> bool:
    return left == right


def main() -> int:
    errors: list[str] = []
    contract = load_yaml(CONTRACT_PATH)
    fields_core = load_yaml(FIELDS_CORE_PATH)
    frontmatter_profiles = load_yaml(FRONTMATTER_PROFILES_PATH)
    runtime_profiles = load_yaml(RUNTIME_PROFILES_PATH)
    allowed_categories = values_for_core_field(fields_core, "categoria")
    allowed_states = values_for_core_field(fields_core, "stato")
    known_fields = known_frontmatter_fields(fields_core, frontmatter_profiles)

    for key in ["live_entity_categories", "codex_categories"]:
        validate_category_list(errors, require_non_empty_array(errors, contract.get(key), key), allowed_categories, key)

    category_validation = require_non_empty_map(errors, contract.get("category_validation"), "category_validation")
    type_exempt_categories = set(
        require_non_empty_array(
            errors,
            category_validation.get("type_exempt_categories"),
            "category_validation.type_exempt_categories",
        )
    )
    required_field_exempt_categories = set(
        require_non_empty_array(
            errors,
            category_validation.get("required_field_exempt_categories"),
            "category_validation.required_field_exempt_categories",
        )
    )
    validate_category_list(
        errors,
        list(type_exempt_categories),
        allowed_categories,
        "category_validation.type_exempt_categories",
    )
    validate_category_list(
        errors,
        list(required_field_exempt_categories),
        allowed_categories,
        "category_validation.required_field_exempt_categories",
    )

    allowed_types_by_category = require_non_empty_map(
        errors,
        contract.get("allowed_types_by_category"),
        "allowed_types_by_category",
    )
    for category, types in allowed_types_by_category.items():
        if category not in allowed_categories:
            errors.append(
                f"{CONTRACT_PATH.relative_to(ROOT)}: allowed_types_by_category contiene categoria "
                f"non dichiarata in fields_core.yaml ({category})"
            )
        require_non_empty_array(errors, types, f"allowed_types_by_category.{category}")
        if category in type_exempt_categories:
            errors.append(
                f"{CONTRACT_PATH.relative_to(ROOT)}: {category} risulta sia in allowed_types_by_category "
                "sia in category_validation.type_exempt_categories"
            )
    for category in allowed_categories:
        if category not in allowed_types_by_category and category not in type_exempt_categories:
            errors.append(
                f"{CONTRACT_PATH.relative_to(ROOT)}: allowed_types_by_category non copre categoria "
                f"{category} e non la dichiara esente"
            )

    runtime_location_types = {
        str(option.get("id") or "")
        for option in (((runtime_profiles.get("profiles") or {}).get("luogo") or {}).get("type_options") or [])
        if isinstance(option, dict) and option.get("id")
    }
    if runtime_location_types:
        contract_location_types = {str(value) for value in allowed_types_by_category.get("luogo", [])}
        if not set_equals(runtime_location_types, contract_location_types):
            errors.append(
                f"{CONTRACT_PATH.relative_to(ROOT)}: allowed_types_by_category.luogo deve corrispondere "
                "a runtime_profiles.luogo.type_options"
            )

    required_fields_by_category = require_non_empty_map(
        errors,
        contract.get("required_fields_by_category"),
        "required_fields_by_category",
    )
    for category, fields in required_fields_by_category.items():
        if category not in allowed_categories:
            errors.append(
                f"{CONTRACT_PATH.relative_to(ROOT)}: required_fields_by_category contiene categoria "
                f"non dichiarata in fields_core.yaml ({category})"
            )
        require_known_field_array(errors, known_fields, fields, f"required_fields_by_category.{category}")
        if category in required_field_exempt_categories:
            errors.append(
                f"{CONTRACT_PATH.relative_to(ROOT)}: {category} risulta sia in required_fields_by_category "
                "sia in category_validation.required_field_exempt_categories"
            )
    for category in allowed_categories:
        if category not in required_fields_by_category and category not in required_field_exempt_categories:
            errors.append(
                f"{CONTRACT_PATH.relative_to(ROOT)}: required_fields_by_category non copre categoria "
                f"{category} e non la dichiara esente"
            )

    ready_requirements = require_non_empty_map(errors, contract.get("state_ready_requirements"), "state_ready_requirements")
    ready_states = require_non_empty_map(errors, ready_requirements.get("pronto"), "state_ready_requirements.pronto")
    allowed_ready_keys = set(allowed_categories)
    allowed_ready_keys.update(str(key) for key in allowed_types_by_category.keys())
    allowed_ready_keys.update(str(value) for values in allowed_types_by_category.values() for value in values)
    for category_or_type, requirement in ready_states.items():
        if category_or_type not in allowed_ready_keys:
            errors.append(
                f"{CONTRACT_PATH.relative_to(ROOT)}: state_ready_requirements.pronto contiene chiave "
                f"non riconducibile a categoria o tipo ({category_or_type})"
            )
        requirement_map = requirement if isinstance(requirement, dict) else {}
        require_known_field_array(
            errors,
            known_fields,
            requirement_map.get("any_of_fields"),
            f"state_ready_requirements.pronto.{category_or_type}.any_of_fields",
        )

    require_known_field_array(errors, known_fields, contract.get("private_frontmatter_fields"), "private_frontmatter_fields")
    require_non_empty_array(errors, contract.get("private_text_terms"), "private_text_terms")

    live_entity_policy = require_non_empty_map(errors, contract.get("live_entity_policy"), "live_entity_policy")
    require_known_field_array(
        errors,
        known_fields,
        live_entity_policy.get("require_any_of_fields"),
        "live_entity_policy.require_any_of_fields",
    )
    require_known_field_array(
        errors,
        known_fields,
        live_entity_policy.get("sheet_require_any_of_fields"),
        "live_entity_policy.sheet_require_any_of_fields",
    )

    codex_article_policy = require_non_empty_map(errors, contract.get("codex_article_policy"), "codex_article_policy")
    for key in [
        "identity_any_of_fields",
        "table_use_any_of_fields",
        "dm_layer_any_of_fields",
        "operational_link_fields",
    ]:
        require_known_field_array(errors, known_fields, codex_article_policy.get(key), f"codex_article_policy.{key}")

    session_playability_policy = require_non_empty_map(
        errors,
        contract.get("session_playability_policy"),
        "session_playability_policy",
    )
    for state in require_non_empty_array(
        errors,
        session_playability_policy.get("active_states"),
        "session_playability_policy.active_states",
    ):
        if state not in allowed_states:
            errors.append(
                f"{CONTRACT_PATH.relative_to(ROOT)}: session_playability_policy.active_states usa stato "
                f"non dichiarato in fields_core.yaml ({state})"
            )
    require_known_field_array(
        errors,
        known_fields,
        session_playability_policy.get("playable_any_of_fields"),
        "session_playability_policy.playable_any_of_fields",
    )
    min_world_anchors = require_number_at_least(
        errors,
        session_playability_policy.get("min_world_anchors"),
        "session_playability_policy.min_world_anchors",
        1,
    )
    world_anchor_groups = require_known_field_groups(
        errors,
        known_fields,
        session_playability_policy.get("world_anchor_groups"),
        "session_playability_policy.world_anchor_groups",
    )
    if world_anchor_groups and min_world_anchors > len(world_anchor_groups):
        errors.append(
            f"{CONTRACT_PATH.relative_to(ROOT)}: session_playability_policy.min_world_anchors "
            "supera i gruppi world_anchor_groups"
        )
    encounter_material_field = require_non_empty_string(
        errors,
        session_playability_policy.get("encounter_material_field"),
        "session_playability_policy.encounter_material_field",
    )
    if encounter_material_field:
        require_known_field(
            errors,
            known_fields,
            encounter_material_field,
            "session_playability_policy.encounter_material_field",
        )

    map_review_policy = require_non_empty_map(errors, contract.get("map_review_policy"), "map_review_policy")
    playable_map_uses = set(
        require_non_empty_array(errors, map_review_policy.get("playable_uses"), "map_review_policy.playable_uses")
    )
    require_non_empty_array(errors, map_review_policy.get("structured_uses"), "map_review_policy.structured_uses")
    for key in [
        "table_use_any_of_fields",
        "visibility_any_of_fields",
        "dm_private_any_of_fields",
        "ready_playable_link_any_of_fields",
        "ready_structural_link_any_of_fields",
    ]:
        require_known_field_array(errors, known_fields, map_review_policy.get(key), f"map_review_policy.{key}")
    zoom_use = require_non_empty_string(errors, map_review_policy.get("zoom_use"), "map_review_policy.zoom_use")
    if zoom_use and zoom_use not in playable_map_uses:
        errors.append(
            f"{CONTRACT_PATH.relative_to(ROOT)}: map_review_policy.zoom_use deve essere incluso "
            "in map_review_policy.playable_uses"
        )

    require_non_empty_array(errors, contract.get("playability_rules"), "playability_rules")
    playability_rules = contract.get("playability_rules") if isinstance(contract.get("playability_rules"), list) else []
    rule_ids: set[str] = set()
    for rule in playability_rules:
        rule_map = rule if isinstance(rule, dict) else {}
        rule_id = str(rule_map.get("id") or "")
        if not rule_id:
            errors.append(f"{CONTRACT_PATH.relative_to(ROOT)}: playability_rules contiene regola senza id")
            continue
        if rule_id in rule_ids:
            errors.append(f"{CONTRACT_PATH.relative_to(ROOT)}: playability_rules id duplicato ({rule_id})")
        rule_ids.add(rule_id)

        if not rule_map.get("warning"):
            errors.append(f"{CONTRACT_PATH.relative_to(ROOT)}: playability_rules.{rule_id} senza warning")
        if not rule_map.get("require_value") and not rule_map.get("require_any_of"):
            errors.append(
                f"{CONTRACT_PATH.relative_to(ROOT)}: playability_rules.{rule_id} "
                "senza require_value o require_any_of"
            )
        if rule_map.get("require_value"):
            require_known_field(errors, known_fields, str(rule_map.get("require_value")), f"playability_rules.{rule_id}.require_value")
        if rule_map.get("require_any_of"):
            require_known_field_array(errors, known_fields, rule_map.get("require_any_of"), f"playability_rules.{rule_id}.require_any_of")
        if rule_map.get("tipo_in"):
            require_non_empty_array(errors, rule_map.get("tipo_in"), f"playability_rules.{rule_id}.tipo_in")
        if rule_map.get("path_prefixes"):
            require_non_empty_array(errors, rule_map.get("path_prefixes"), f"playability_rules.{rule_id}.path_prefixes")
        if rule_map.get("when_value_present"):
            require_known_field(
                errors,
                known_fields,
                str(rule_map.get("when_value_present")),
                f"playability_rules.{rule_id}.when_value_present",
            )
        if rule_map.get("when_any_of_present"):
            require_known_field_array(
                errors,
                known_fields,
                rule_map.get("when_any_of_present"),
                f"playability_rules.{rule_id}.when_any_of_present",
            )
        if rule_map.get("any_field_equals"):
            require_non_empty_array(errors, rule_map.get("any_field_equals"), f"playability_rules.{rule_id}.any_field_equals")
        for condition in rule_map.get("any_field_equals", []) or []:
            condition_map = condition if isinstance(condition, dict) else {}
            require_known_field(
                errors,
                known_fields,
                str(condition_map.get("field") or ""),
                f"playability_rules.{rule_id}.any_field_equals.field",
            )
        if rule_map.get("number_field_gt"):
            condition_map = rule_map.get("number_field_gt") if isinstance(rule_map.get("number_field_gt"), dict) else {}
            require_known_field(
                errors,
                known_fields,
                str(condition_map.get("field") or ""),
                f"playability_rules.{rule_id}.number_field_gt.field",
            )
        if rule_map.get("number_field_gte"):
            condition_map = rule_map.get("number_field_gte") if isinstance(rule_map.get("number_field_gte"), dict) else {}
            require_known_field(
                errors,
                known_fields,
                str(condition_map.get("field") or ""),
                f"playability_rules.{rule_id}.number_field_gte.field",
            )
        if rule_map.get("categoria") and str(rule_map.get("categoria")) not in allowed_categories:
            errors.append(
                f"{CONTRACT_PATH.relative_to(ROOT)}: playability_rules.{rule_id} usa categoria "
                f"non dichiarata in fields_core.yaml ({rule_map.get('categoria')})"
            )

    if errors:
        print("Validation contract non valido:", file=sys.stderr)
        for error in errors:
            print(f"- {error}", file=sys.stderr)
        return 1

    print(
        "Validation contract OK: "
        f"{len(allowed_types_by_category)} categorie tipo, "
        f"{len(required_fields_by_category)} categorie required, "
        f"{len(ready_states)} policy pronto, "
        f"{len(playability_rules)} regole giocabilita, policy live/sessione/mappe validate."
    )
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
